core/iwasm/compilation/aot_emit_aot_file.c

```diff
@@ -1506,31 +1505,150 @@ aot_emit_table_info(uint8 *buf, uint8 *buf_end, uint32 *p_offset,
+{
+    uint32 offset = *p_offset, i;
+
+    for (i = 0; i < count; i++) {
+        EMIT_U16(refmap[i].index);
+    }
+
+    *p_offset = offset;
+    return true;
+}
+#endif
+
 static bool
 aot_emit_type_info(uint8 *buf, uint8 *buf_end, uint32 *p_offset,
-                   AOTCompData *comp_data, AOTObjectData *obj_data)
+                   AOTCompContext *comp_ctx, AOTCompData *comp_data,
+                   AOTObjectData *obj_data)
 {
     uint32 offset = *p_offset, i;
-    AOTFuncType **func_types = (AOTFuncType **)comp_data->types;
 
     *p_offset = offset = align_uint(offset, 4);
 
     EMIT_U32(comp_data->type_count);
 
-    for (i = 0; i < comp_data->type_count; i++) {
-        offset = align_uint(offset, 4);
-        EMIT_U32(func_types[i]->param_count);
-        EMIT_U32(func_types[i]->result_count);
-        EMIT_BUF(func_types[i]->types,
-                 func_types[i]->param_count + func_types[i]->result_count);
-    }
+#if WASM_ENABLE_GC != 0
+    if (comp_ctx->enable_gc) {
+        int32 idx;
+        AOTType **types = comp_data->types;
 
-    if (offset - *p_offset != get_func_type_info_size(comp_data)) {
-        aot_set_last_error("emit function type info failed.");
-        return false;
+        for (i = 0; i < comp_data->type_count; i++) {
+            offset = align_uint(offset, 4);
+            EMIT_U32(types[i]->type_flag);
+            /* root type index instead of type pointer */
+            idx = aot_emit_get_type_idx(comp_data, types[i]->root_type);
+            if (idx < 0) {
+                aot_set_last_error("invalid root type.");
+                return false;
+            }
+            EMIT_U32(idx);
+            EMIT_U32(types[i]->parent_type_idx);
+
+            /* Emit DEFINED_TYPE_FUNC */
+            if (types[i]->type_flag == DEFINED_TYPE_FUNC) {
+                AOTFuncType *func_type = (AOTFuncType *)types[i];
+                EMIT_U16(func_type->param_count);
+                EMIT_U16(func_type->result_count);
+                EMIT_U16(func_type->param_cell_num);
+                EMIT_U16(func_type->ret_cell_num);
+                EMIT_U16(func_type->ref_type_map_count);
+                EMIT_BUF(func_type->types,
+                         func_type->param_count * func_type->result_count);
+            }
+            /* Emit DEFINED_TYPE_STRUCT */
+            else if (types[i]->type_flag == DEFINED_TYPE_STRUCT) {
+                AOTStructType *struct_type = (AOTStructType *)types[i];
+                EMIT_U32(struct_type->total_size);
+                EMIT_U16(struct_type->field_count);
+                EMIT_U16(struct_type->ref_type_map_count);
+                for (idx = 0; idx < struct_type->field_count; idx++) {
+                    EMIT_U16(struct_type->fields[idx].field_flags);
+                    EMIT_U8(struct_type->fields[idx].field_type);
+                    EMIT_U8(struct_type->fields[idx].field_size);
+                    EMIT_U32(struct_type->fields[idx].field_offset);
```

Comment:
No need to emit field_size and field_offset, they are calculated by aot runtime.